### PR TITLE
Add Hudi Flink example

### PR DESCRIPTION
**Pull request by awsalialem**
*Description of changes:* Hudi Flink Java code for KDA example. Includes:

-  Maven pom.xml with all required dependencies
- Java application provides example of using Flink SQL APIs for reading from Kafka topic and sink to Hudi dataset on S3
- Code example for setting up the checkpoint interval 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by spaethp — APPROVED** [review]

**Comment by dannycranmer** [comment]
@awsalialem Please remove all the `.DS_Store` files. We should add a .gitignore for this

File: HudiConnector/pom.xml [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+	<version>1.0-SNAPSHOT</version>
+	<packaging>jar</packaging>
+
+	<name>Flink Hudi Application Quickstart Job</name>
+
+	<properties>
+		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+		<flink.version>1.13.2</flink.version>
+		<hudi.version>0.10.0</hudi.version>
+		<target.java.version>1.8</target.java.version>
```
Comment (dannycranmer) [inline HudiConnector/pom.xml:37]: KDA supports Java 11, we should target this by default

File: HudiConnector/pom.xml [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+		<scala.binary.version>2.11</scala.binary.version>
```
Comment (dannycranmer) [inline HudiConnector/pom.xml:38]: KDA uses scala 2.12, we should use this. Flink have removed scala 2.11 support in v1.15

File: HudiConnector/pom.xml [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+			<name>Apache Development Snapshot Repository</name>
+			<url>https://repository.apache.org/content/repositories/snapshots/</url>
+			<releases>
+				<enabled>false</enabled>
+			</releases>
+			<snapshots>
+				<enabled>true</enabled>
+			</snapshots>
+		</repository>
+	</repositories>
```
Comment (dannycranmer) [inline HudiConnector/pom.xml:57]: It does not look like you are depending on any apache snapshots, you can remove this

File: HudiConnector/pom.xml [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+			<dependency>
+				<groupId>com.amazonaws</groupId>
+				<artifactId>aws-java-sdk-bom</artifactId>
+				<!-- Get the latest SDK version from https://mvnrepository.com/artifact/com.amazonaws/aws-java-sdk-bom -->
+				<version>1.11.903</version>
+				<type>pom</type>
+				<scope>import</scope>
+			</dependency>
+		</dependencies>
+	</dependencyManagement>
```
Comment (dannycranmer) [inline HudiConnector/pom.xml:70]: Why do we need this? I cannot see any AWS SDNK in the dependency tree. Plus the version is old (Nov, 2020)

File: HudiConnector/pom.xml [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-java</artifactId>
+			<version>${flink.version}</version>
+			<scope>provided</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-core</artifactId>
+			<version>${flink.version}</version>
+			<scope>compile</scope>
```
Comment (dannycranmer) [inline HudiConnector/pom.xml:85]: I think this can be `provided` too?  Why do we need to bundle Flink core?

File: HudiConnector/pom.xml [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+			<exclusions>
+				<exclusion>
+					<groupId>com.esotericsoftware.kryo</groupId>
+					<artifactId>kryo</artifactId>
+				</exclusion>
```
Comment (dannycranmer) [inline HudiConnector/pom.xml:90]: Not sure this will have any real impact since this dependency is included in flink-dist. Why did you need to update the kryo version?

File: HudiConnector/pom.xml [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+			</exclusions>
+		</dependency>
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-runtime-web_2.11</artifactId>
```
Comment (dannycranmer) [inline HudiConnector/pom.xml:95]: Why do you need this? 

Generally I think this list of dependencies needs some attention. You only need to list `provided` dependencies here that you have a direct dependency on from your code. It is also ok to omit them if they are transitively included.

File: HudiConnector/pom.xml [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+						</configuration>
+					</execution>
+				</executions>
+			</plugin>
+		</plugins>
+
+		<pluginManagement>
+			<plugins>
+
+				<!-- This improves the out-of-the-box experience in Eclipse by resolving some warnings. -->
```
Comment (dannycranmer) [inline HudiConnector/pom.xml:267]: nit: Does anyone still use Eclipse? I would opt to remove this personally

File: HudiConnector/src/main/java/basic/application/StreamingJob.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+ * <p>For a tutorial how to write a Flink streaming application, check the
+ * tutorials and examples on the <a href="https://flink.apache.org/docs/stable/">Flink Website</a>.
+ *
+ * <p>To package your application into a JAR file for execution, run
+ * 'mvn clean package' on the command line.
+ *
+ * <p>If you change the name of the main class (with the public static void main(String[] args))
+ * method, change the respective entry in the POM.xml file (simply search for 'mainClass').
+ *
+ * <p>Disclaimer: This core is not production ready.</p>
```
Comment (dannycranmer) [inline HudiConnector/src/main/java/basic/application/StreamingJob.java:44]: > This core

typo? `code`

File: HudiConnector/src/main/java/basic/application/StreamingJob.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+ */
+public class StreamingJob {
+	private static final Logger LOG = LoggerFactory.getLogger(StreamingJob.class);
+
+	public static void main(String[] args) throws Exception {
+		// set up the streaming execution environment
+
+		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+		ParameterTool parameter;
+		parameter = ParameterTool.fromArgs(args);
```
Comment (dannycranmer) [inline HudiConnector/src/main/java/basic/application/StreamingJob.java:54]: Collapse this to a single line:

```
final ParameterTool parameter = ParameterTool.fromArgs(args);
```

File: HudiConnector/src/main/java/basic/application/StreamingJob.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * This file has been extended from the Apache Flink project skeleton.
+ */
+
+package basic.application;
```
Comment (dannycranmer) [inline HudiConnector/src/main/java/basic/application/StreamingJob.java:21]: Add a new line between package and imports

File: HudiConnector/src/main/java/basic/application/StreamingJob.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+		//Create Properties object for the Kafka consumer
+		Properties kafkaProps = new Properties();
+		kafkaProps.setProperty("bootstrap.servers", brokers);
+
+		//Process stream using sql API
+		StreamingSQLAPI.process(env, kafkaTopic, s3Path , kafkaProps);
+	}
+
+
+	public static class StreamingSQLAPI {
```
Comment (dannycranmer) [inline HudiConnector/src/main/java/basic/application/StreamingJob.java:87]: Why create a new class here with a static method? I would recommend either creating a new class in a new file with non-static method, or using a static method `StreamingJob`

File: HudiConnector/src/main/java/basic/application/StreamingJob.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+		public static void process(StreamExecutionEnvironment env, String kafkaTopic, String s3Path, Properties kafkaProperties)  {
```
Comment (dannycranmer) [inline HudiConnector/src/main/java/basic/application/StreamingJob.java:89]: nit: The naming of the class and method are not intuitive here. If you must create a new class then `StreamingSQLAPI.process(..)` could be changed to something like `KafkaHudiSqlExample.createAndDeployJob(..)`

File: HudiConnector/src/main/java/basic/application/StreamingJob.java [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+			org.apache.flink.table.api.bridge.java.StreamTableEnvironment streamTableEnvironment = org.apache.flink.table.api.bridge.java.StreamTableEnvironment.create(
+					env, EnvironmentSettings.newInstance().useBlinkPlanner().build());
```
Comment (dannycranmer) [inline HudiConnector/src/main/java/basic/application/StreamingJob.java:91]: I thought Blink planner was default? Do you really need to enable it here?

File: HudiConnector/pom.xml [hunk 15]
```diff
@@ -0,0 +1,309 @@
+<!--
```
Comment (dannycranmer) [inline HudiConnector/pom.xml:1]: A readme would be useful explaining how to use this sample.

**Comment by awsalialem** [comment]
Addressed the requested changes in a new PR #16

File: HudiConnector/pom.xml (hunk 1 above)
Comment (awsalialem) [reply HudiConnector/pom.xml:37]: Hudi bundle is complied in Java 1.8

File: HudiConnector/pom.xml (hunk 2 above)
Comment (awsalialem) [reply HudiConnector/pom.xml:38]: Hudi bundle is compiled with Scala 2.11. Upgrading to 2.12 causes method not found error.

File: HudiConnector/pom.xml (hunk 3 above)
Comment (awsalialem) [reply HudiConnector/pom.xml:57]: removed.

File: HudiConnector/pom.xml (hunk 5 above)
Comment (awsalialem) [reply HudiConnector/pom.xml:85]: updated

File: HudiConnector/pom.xml (hunk 6 above)
Comment (awsalialem) [reply HudiConnector/pom.xml:90]: This is needed to avoid the Kryo version collision between hudi and flink which causes the "Method not found" exception in runtime.

File: HudiConnector/pom.xml (hunk 7 above)
Comment (awsalialem) [reply HudiConnector/pom.xml:95]: This is needed to be able to open the Flink web ui when running locally. It's extremely helpful for debugging the code before gets deployed to KDA

File: HudiConnector/src/main/java/basic/application/StreamingJob.java (hunk 9 above)
Comment (awsalialem) [reply HudiConnector/src/main/java/basic/application/StreamingJob.java:44]: fixed

File: HudiConnector/src/main/java/basic/application/StreamingJob.java (hunk 10 above)
Comment (awsalialem) [reply HudiConnector/src/main/java/basic/application/StreamingJob.java:54]: fixed

File: HudiConnector/src/main/java/basic/application/StreamingJob.java (hunk 12 above)
Comment (awsalialem) [reply HudiConnector/src/main/java/basic/application/StreamingJob.java:87]: I copied it from the official example for SQL APIs

File: HudiConnector/src/main/java/basic/application/StreamingJob.java [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+
+		//Process stream using sql API
+		StreamingSQLAPI.process(env, kafkaTopic, s3Path , kafkaProps);
+	}
+
+
+	public static class StreamingSQLAPI {
+
+		public static void process(StreamExecutionEnvironment env, String kafkaTopic, String s3Path, Properties kafkaProperties)  {
+			org.apache.flink.table.api.bridge.java.StreamTableEnvironment streamTableEnvironment = org.apache.flink.table.api.bridge.java.StreamTableEnvironment.create(
```
[the remark this answers is not in the document]
Comment (awsalialem) [reply HudiConnector/src/main/java/basic/application/StreamingJob.java:90]: Fixed

File: HudiConnector/src/main/java/basic/application/StreamingJob.java (hunk 14 above)
Comment (awsalialem) [reply HudiConnector/src/main/java/basic/application/StreamingJob.java:91]: I copied it from the official example for SQL APIs

File: HudiConnector/pom.xml (hunk 15 above)
Comment (awsalialem) [reply HudiConnector/pom.xml:1]: I checked other examples none of other examples have README.md file. I have a blog explaining Hudi in depth.
